include/RED4ext/ResourceReference.hpp

```diff
@@ -0,0 +1,132 @@
+#pragma once
+
+#include <cstdint>
+
+#include <RED4ext/Common.hpp>
+#include <RED4ext/Handle.hpp>
+#include <RED4ext/ResourcePath.hpp>
+#include <RED4ext/Scripting/Natives/Generated/CResource.hpp>
+
+namespace RED4ext
+{
+struct ResourceReference
+{
+    ResourceReference() noexcept = default;
+    ResourceReference(ResourcePath aPath) noexcept;
+    ResourceReference(ResourcePath aPath, bool aLoad) noexcept;
+    ResourceReference(const ResourceReference& aOther) noexcept;
+    ResourceReference(ResourceReference&& aOther) noexcept;
+    ~ResourceReference();
+
+    /**
+     * @brief Load the resource synchronously and fill the token.
+     */
+    void Load();
+
+    /**
+     * @brief Load if it's not already loaded and get the resource.
+     * @return The loaded resource.
+     */
+    Handle<CResource>& Fetch();
+
+    /**
+     * @brief Reset the path and associated token.
+     */
+    void Reset();
+
+    struct ResourceTokenPtr
+    {
+        void* instance{ nullptr };
+        RED4ext::RefCnt* refCount{ nullptr };
+    };
+
+    ResourcePath path;      // 00
+    ResourceTokenPtr token; // 08
+};
+RED4EXT_ASSERT_SIZE(ResourceReference, 0x18);
+
+struct ResourceAsyncReference
+{
+    ResourceAsyncReference() noexcept = default;
+    ResourceAsyncReference(ResourcePath aPath) noexcept;
+    ResourceAsyncReference(const ResourceAsyncReference& aOther) noexcept;
+    ResourceAsyncReference(ResourceAsyncReference&& aOther) noexcept;
+
+    [[nodiscard]] ResourceReference Resolve() const noexcept;
+
+    ResourcePath path; // 00
+};
+RED4EXT_ASSERT_SIZE(ResourceAsyncReference, 0x8);
+
+template<typename T = CResource>
+struct TResourceReference : ResourceReference
+{
+    // static_assert(std::is_base_of_v<CResource, T> or std::is_same_v<CResource, T>,
```

Comment:
Should this be commented?

Reply:
Oh right... The assertion is correct, but it cannot be used with current generated classes.
For example, `include/RED4ext/Scripting/Natives/Generated/AI/ArchetypeSetEntry.hpp` defines:
```c++
namespace AI { struct Archetype; }
...
Ref<AI::Archetype> resource;
```
And the assertion fails even though the real `AI::Archetype` is `CResource`.